geps/gep-1713.md

```diff
@@ -0,0 +1,220 @@
+  // A common scenario is to reference another Gateway which marks
+  // this Gateway a child of another.
+  AttachTo LocalObjectReference `json:"attachTo"`
+}
+```
+
+#### YAML
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: child-gateway
+spec:
+  gatewayClassName: example
+  infrastructure:
+    attachTo:
+      name: parent-gateway
+      kind: Gateway
+      group: gateway.networking.k8s.io
+  listeners:
+  - name: metrics
+    port: 8080
+    protocol: HTTP
+```
+
+### Semantics
+
+#### Gateway Attaching
+
+Gateways that _do not_ reference another Gateway (using `spec.infrastructure.attachTo`) are considered "primary" Gateways. A "child" Gateway is one where `spec.infrastructure.attachTo` field is set to a "primary" Gateway - making the "primary" Gateway a "parent". A "sibling" is a Gateway that shares a "parent" with another "child" Gateway. A "child" Gateway can only attach to Gateways in the same namespace.
+
+Failure to attach a "child" Gateway to another MUST result in the "child" Gateway setting their `Accepted` condition to `False`.
+
+This MUST happen in the following scenarios:
+- Attaching to a non-existing Gateway
+
+If a Gateway's `Accepted` condition is `False` then the conditions 'Programmed' and 'Ready' (if supported) MUST be set to `False`.
+
+Invalid Gateways MUST be rejected by the implementation during creation or update. This MUST happen in the following scenarios:
+- Attaching to another "child" Gateway
+- Attaching to oneself (self-referential)
+- Attaching to a Gateway with a different `GatewayClassName`
+
+#### Route Attaching
+
+Routes MUST be able to specify a "child" Gateways as a `parentRef` and make use of the fields in `ParentReference` to help target a specific listener.
+
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: httproute-example
+spec:
+  parentRefs:
+  - name: child-gateway
+    sectionName: metrics
+```
+
+#### Merging Spec Fields
+
+##### GatewayClassName
+
+The `GatewayClassName` MUST be set on "child" and "parent" Gateways and be equivalent.
+
+##### Listeners
+
+Implementations MUST treat the "parent" Gateway as having the concatenated list of all listeners from itself and "child" Gateways.
+
+Validation of this list of listeners is subject to the discussion here: https://github.com/kubernetes-sigs/gateway-api/issues/1842
+
+eg.
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: parent-gateway
+spec:
+  gatewayClassName: example
+  listeners:
+  - name: HTTP
+    port: 80
+    protocol: HTTP 
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: child-gateway
+spec:
+  gatewayClassName: example
+  infrastructure:
+    attachTo:
+      name: parent-gateway
+      kind: Gateway
+      group: gateway.networking.k8s.io
+  listeners:
+  - name: metrics
+    port: 8080
+    protocol: HTTP    
```

Comment:
How would you see a listener with the same name as one in the parent being merged (or not).
For example, in the parent:

```yaml
    - allowedRoutes:
        namespaces:
          from: All
      hostname: test.example.com
      name: api
      port: 443
      protocol: HTTPS
```
and in the child:

```yaml
    - allowedRoutes:
        namespaces:
          from: All
      hostname: test.example.com
      name: api
      port: 443
      protocol: HTTPS
      tls:
        certificateRefs:
        - group: ""
          kind: Secret
          name: test.example.com
          namespace: example
        mode: Terminate
```

The use case here is a parent Gateway managed by a developer, and child gateways that are automatically created & managed by a controller (in this case supplementing the config with a tls secret reference that my controller is managing).

Reply:
> How would you see a listener with the same name as one in the parent being merged (or not).

You need distinct names as per the merging logic defined here - https://github.com/kubernetes-sigs/gateway-api/blob/f2b55c8eaf833163f751ff1b9b2abe61db8e7343/apis/v1beta1/gateway_types.go#L87


> The use case here is a parent Gateway managed by a developer, and child gateways that are automatically created & managed by a controller (in this case supplementing the config with a tls secret reference that my controller is managing).

That's exactly the use case I'm hoping to cover. Though you'll need a distinct listener name and host name as per the link I sent.

This GEP doesn't adjust the merge semantics of a list of listeners. This GEP is really just splits that list of listeners over N objects so they can be manage more easily (eg. creating and reading status etc)

Reply:
I realize I left a lingering link in the GEP let me update it

Reply:
done